### PR TITLE
JA-9 JA-10: OAuth2 and JWT Authorization and Authentication

### PR DESCRIPTION
**Pull request by kentherussell**
Please pull this down and try before approving. There are environment variables that need to be set and a JKS file that needs to be generated. Just need to make sure instructions are thorough enough to setup correctly.

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -13,7 +13,7 @@ pipeline {
         stage('Unit Tests') {
             steps {
                 sh 'java -version'
-                gradlew('test')
+                gradlew('testNG')
```
Comment (vickumar1981) [inline Jenkinsfile]: full command to run linter, tests, and check coverage reports: `verGJF clean testNG jacocoTestReport jacocoTestCoverageVerification` - https://github.com/excellaco/tcp-java/blob/master/test#L1

File: README.md [hunk 2]
```diff
@@ -5,7 +5,45 @@ A CRUD example using Spring WebFlux and Java 11
 
 ##### 1. Install [docker](https://docs.docker.com/engine/installation/) and [docker-compose](https://docs.docker.com/compose/install/)
 
-##### 2. Running the application
+
+##### 2. Setup the application
```
Comment (vickumar1981) [inline README.md]: :+1:  i followed these instructions and it worked fine, so the config looks good.

File: src/main/java/com/excella/reactor/common/exceptions/CustomSetupException.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package com.excella.reactor.common.exceptions;
+
+public class CustomSetupException extends RuntimeException {
```
Comment (vickumar1981) [inline src/main/java/com/excella/reactor/common/exceptions/CustomSetupException.java]: is this class used?

File: src/main/java/com/excella/reactor/config/AuthorizationConfig.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+   * and the Authentication Manager.
+   *
+   * @param tokenStore tokenStore
+   * @param clientDetailsService clientDetailsService
+   * @return Default token service
+   */
+  @Bean
+  public DefaultTokenServices tokenServices(
+      final TokenStore tokenStore, final ClientDetailsService clientDetailsService) {
+    DefaultTokenServices tokenServices = new DefaultTokenServices();
```
Comment (vickumar1981) [inline src/main/java/com/excella/reactor/config/AuthorizationConfig.java]: could use `var` here.

File: src/main/resources/db/migration/V3__OAuthTables.sql [hunk 5]
```diff
@@ -0,0 +1,57 @@
+create table oauth_client_details
```
Comment (vickumar1981) [inline src/main/resources/db/migration/V3__OAuthTables.sql]: just curious, but is there a link to the migrations needed to setup spring's oauth?  might be helpful to link to somewhere in the `README.md`.

File: README.md (hunk 2 above)
Comment (vickumar1981) [inline README.md]: Tried the instructions, but I'm not sure if the environment variables are getting set inside the docker container - I'm getting these warnings when running `./start bootrun`.

```
WARNING: The JWT_KEY_STORE variable is not set. Defaulting to a blank string.
WARNING: The JWT_KEY_STORE_PASS variable is not set. Defaulting to a blank string.
WARNING: The JWT_KEYPAIR_ALIAS variable is not set. Defaulting to a blank string.
WARNING: The JWT_KEYPAIR_PASS variable is not set. Defaulting to a blank string.
```

I'm using `Docker version 18.09.6, build 481bc77`

File: README.md [hunk 6]
```diff
@@ -5,7 +5,45 @@ A CRUD example using Spring WebFlux and Java 11
...
+                    -storepass mypass
+```
+Note: it's recommended you use the values in the application.yml file. 
+
+Next, add an .env file and populate it with the same info you specified in the `keytool` command above.
+
+Here's what your .env file should look like.
+
+```$xslt
+KEY_STORE=classpath:server.jks
```
Comment (vickumar1981) [inline README.md]: I ** think **, or at least for me, I had to add `JWT_` in front of these.

File: Jenkinsfile (hunk 1 above)
Comment (kentherussell) [reply Jenkinsfile]: Should we separate that into different steps in the Jenkins file (verification, testing, coverage)?

File: src/main/java/com/excella/reactor/common/exceptions/CustomSetupException.java (hunk 3 above)
Comment (kentherussell) [reply src/main/java/com/excella/reactor/common/exceptions/CustomSetupException.java]: Didn't realize that was still there. Deleted

File: src/main/java/com/excella/reactor/config/AuthorizationConfig.java (hunk 4 above)
Comment (kentherussell) [reply src/main/java/com/excella/reactor/config/AuthorizationConfig.java]: good point

File: src/main/resources/db/migration/V3__OAuthTables.sql (hunk 5 above)
Comment (kentherussell) [reply src/main/resources/db/migration/V3__OAuthTables.sql]: That's a good point. That's definitely something to add in the README

File: README.md (hunk 6 above)
Comment (kentherussell) [reply README.md]: I just modified this. I switched over to a `JWT_` prefix to signify why we need the JKS file.

File: README.md (hunk 2 above)
Comment (kentherussell) [reply README.md]: Thanks!

File: README.md [hunk 7]
```diff
@@ -5,7 +5,48 @@ A CRUD example using Spring WebFlux and Java 11
...
+```
+Note: it's recommended you use the values in the application.yml file. 
+
+Next, add an .env file and populate it with the same info you specified in the `keytool` command above.
+
+Here's what your .env file should look like.
+
+```$xslt
+JWT_KEY_STORE=classpath:server.jks
+JWT+KEY_STORE_PASS=74Ew4g77XGnFC
```
Comment (vickumar1981) [inline README.md]: type - `_`

File: build.gradle [hunk 8]
```diff
@@ -40,6 +40,11 @@ jacocoTestReport {
 jacocoTestCoverageVerification {
 	violationRules {
 		rule {
+			element = 'CLASS'
+			excludes = [
+					"**/domain/**",
```
Comment (vickumar1981) [inline build.gradle]: i think we ** can **, maybe ** should **, test our domain models.

File: README.md (hunk 7 above)
Comment (kentherussell) [reply README.md]: fixed

File: Jenkinsfile (hunk 1 above)
Comment (vickumar1981) [reply Jenkinsfile]: i think that's a great idea.  1st, i think we need to get this working also.

File: build.gradle (hunk 8 above)
Comment (kentherussell) [reply build.gradle]: Dale and I were just discussing this. I can take it off exclusions

**Pull request merged**